feature: Allow sending static key/value pairs to the configured IdP

strehle opened the pull request:
Feature: Allow sending static key/value pairs to the configured external IdP for authorization code flow

Add parameter additionalAuthzParameters (additional authorization parameters) to requests for auth code flow. The usage is: OIDC IdPs have additional parameters, e.g. UAA itself has token_format to get special features. The new parameter allows to customize the IdP with own parameters

strehle committed:
Feature: Allow sending static key/value pairs to the configured external IdP for authorization code flow

Add parameter additionalAuthzParameters (additional authorization parameters) to requests for auth code flow.
The usage is: OIDC IdPs have additional parameters, e.g. UAA itself has token_format to get special features.
The new parameter allows to customize the IdP with own parameters

strehle committed:
Documentation added

strehle committed:
tests

strehle committed:
more tests

adrianhoelzl-sap commented on model/src/test/java/org/cloudfoundry/identity/uaa/provider/OIDCIdentityProviderDefinitionTests.java:
```diff
@@ -28,11 +28,11 @@
 public class OIDCIdentityProviderDefinitionTests {
 
-    private final String defaultJson = "{\"emailDomain\":null,\"additionalConfiguration\":null,\"providerDescription\":null,\"externalGroupsWhitelist\":[],\"attributeMappings\":{},\"addShadowUserOnLogin\":true,\"storeCustomAttributes\":false,\"authUrl\":null,\"tokenUrl\":null,\"tokenKeyUrl\":null,\"tokenKey\":null,\"linkText\":null,\"showLinkText\":true,\"skipSslValidation\":false,\"relyingPartyI ...
+    private final String defaultJson = "{\"emailDomain\":null,\"additionalConfiguration\":null,\"providerDescription\":null,\"externalGroupsWhitelist\":[],\"attributeMappings\":{},\"addShadowUserOnLogin\":true,\"storeCustomAttributes\":false,\"authUrl\":null,\"tokenUrl\":null,\"tokenKeyUrl\":null,\"tokenKey\":null,\"linkText\":null,\"showLinkText\":true,\"skipSslValidation\":false,\"relyingPartyI ...
```
```suggestion
    private final String defaultJson = "{\"emailDomain\":null,\"additionalConfiguration\":null,\"providerDescription\":null,\"externalGroupsWhitelist\":[],\"attributeMappings\":{},\"addShadowUserOnLogin\":true,\"storeCustomAttributes\":false,\"authUrl\":null,\"tokenUrl\":null,\"tokenKeyUrl\":null,\"tokenKey\":null,\"linkText\":null,\"showLinkText\":true,\"skipSslValidation\":false,\"relyingPartyId\":null,\"relyingPartySecret\":null,\"scopes\":null,\"issuer\":null,\"responseType\":\"code\",\"userInfoUrl\":null,\"jwtClientAuthentication\":false,\"additionalAuthzParameters\":{\"token_format\":\"jwt\"}}";
```

strehle replied:
thanks, fixed

tack-sap commented on model/src/test/java/org/cloudfoundry/identity/uaa/provider/OIDCIdentityProviderDefinitionTests.java:
```diff
@@ -28,11 +28,11 @@
 public class OIDCIdentityProviderDefinitionTests {
 
-    private final String defaultJson = "{\"emailDomain\":null,\"additionalConfiguration\":null,\"providerDescription\":null,\"externalGroupsWhitelist\":[],\"attributeMappings\":{},\"addShadowUserOnLogin\":true,\"storeCustomAttributes\":false,\"authUrl\":null,\"tokenUrl\":null,\"tokenKeyUrl\":null,\"tokenKey\":null,\"linkText\":null,\"showLinkText\":true,\"skipSslValidation\":false,\"relyingPartyI ...
+    private final String defaultJson = "{\"emailDomain\":null,\"additionalConfiguration\":null,\"providerDescription\":null,\"externalGroupsWhitelist\":[],\"attributeMappings\":{},\"addShadowUserOnLogin\":true,\"storeCustomAttributes\":false,\"authUrl\":null,\"tokenUrl\":null,\"tokenKeyUrl\":null,\"tokenKey\":null,\"linkText\":null,\"showLinkText\":true,\"skipSslValidation\":false,\"relyingPartyI ...
```
Typo in json

strehle replied:
thanks, fixed

adrianhoelzl-sap commented on model/src/test/java/org/cloudfoundry/identity/uaa/provider/OIDCIdentityProviderDefinitionTests.java:
```diff
@@ -46,6 +46,14 @@ public void serialize_discovery_url() throws MalformedURLException {
+        OIDCIdentityProviderDefinition def = JsonUtils.readValue(defaultJson, OIDCIdentityProviderDefinition.class);
+        OIDCIdentityProviderDefinition def2 = (OIDCIdentityProviderDefinition) def.clone();
+        assertTrue(def.equals(def2));
+        assertEquals(def.hashCode(), def2.hashCode());
```
Consider adding an assertion for the `additionalAuthzParameters` property. As there was a typo above and the test execution still succeeded, we know that it did not cover your changes.

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthProviderConfigurator.java:
```diff
@@ -92,6 +95,9 @@ public String getIdpAuthenticationUrl(
             uriBuilder.queryParam("nonce", nonceGenerator.generate());
+
+            Map<String, String> additionalParameters = ofNullable(((OIDCIdentityProviderDefinition) definition).getAdditionalAuthzParameters()).orElse(emptyMap());
+            additionalParameters.keySet().stream().filter(Objects::nonNull).forEach(e -> uriBuilder.queryParam(e, additionalParameters.get(e)));
```
Do we really need the `filter` here? `keySet` should not contain `null` values anyway.

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/OauthIDPWrapperFactoryBean.java:
```diff
@@ -142,8 +147,13 @@ protected void setCommonProperties(Map<String, Object> idpDefinitionMap, Abstrac
+                oidcIdentityProviderDefinition.setAdditionalAuthzParameters(parseAdditionalParameters(idpDefinitionMap));
+            }
+            if (hasText(discoveryUrl) && oidcIdentityProviderDefinition != null) {
+                oidcIdentityProviderDefinition.setDiscoveryUrl(new URL(discoveryUrl));
```
```suggestion
            if (idpDefinition instanceof OIDCIdentityProviderDefinition) {
                ((OIDCIdentityProviderDefinition) idpDefinition).setAdditionalAuthzParameters(parseAdditionalParameters(idpDefinitionMap));
                
                if (hasText(discoveryUrl) {
                     ((OIDCIdentityProviderDefinition) idpDefinition).setDiscoveryUrl(new URL(discoveryUrl));
                }
            }
```

adrianhoelzl-sap replied:
In our opinion, using the `oidcIdentityProviderDefinition` variable in the second `if` makes it harder to understand the intended logic.

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/OauthIDPWrapperFactoryBean.java:
```diff
@@ -159,6 +169,29 @@ protected void setCommonProperties(Map<String, Object> idpDefinitionMap, Abstrac
+                    value = (String) entry.getValue();
+                }
+                // accept only custom parameters
+                if (value == null || oauthParameters.contains(keyEntry)) {
```
Do we need a similar check for the identity provider endpoints? There, the reserved property names could also be used.

strehle replied:
this bean loads the data from YAML into internal config, so only allowed parameters are taken over to internal structure and during runtime no need to check again. 

See test https://github.com/cloudfoundry/uaa/pull/2397/files#diff-f941999144a906931d44045832be571ee2874de80265a762d87fd0fc6d87d01dR172-R191

tack-sap replied:
Adding the idp via YAML is only one possible method. One could also add idps directly via the endpoints and then no check will happen on create or update of the idp. So we could end up with parameters that are not allowed via yaml configuration in an idp that was created via the endpoints.
If we do introduce such restrictions on possible values, we should have them in all places.

strehle replied:
Created now a check in REST. Reject invalid settings in REST but filter out parameters in YAML (because of startup issues)

adrianhoelzl-sap commented on server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthProviderConfiguratorTests.java:
```diff
@@ -107,6 +107,7 @@ void setup() throws MalformedURLException {
         config.setRelyingPartySecret("identitysecret");
         config.setResponseType("id_token");
         config.setScopes(List.of("openid", "cloud_controller.read"));
+        config.setAdditionalAuthzParameters(Map.of("token_format", "jwt"));
```
Can you please add a test case to this class that checks whether the additional authorization parameters are added properly to the authentication URLs?

adrianhoelzl-sap commented on server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthAuthenticationManagerTest.java:
```diff
@@ -135,6 +135,7 @@ public void setup() throws Exception {
             entry(GROUP_ATTRIBUTE_NAME, "roles")
         );
         oidcConfig.setAttributeMappings(externalGroupMapping);
+        oidcConfig.setAdditionalAuthzParameters(Map.of("token_format", "jwt"));
```
Can you please add a test case here that checks whether the additional authz parameters are properly added to the call to the IdP?

adrianhoelzl-sap commented on server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthAuthenticationManagerIT.java:
```diff
@@ -263,6 +263,7 @@ void setUp() throws Exception {
                 .setRelyingPartySecret("identitysecret")
                 .setUserInfoUrl(new URL("http://localhost/userinfo"))
                 .setTokenKey(PUBLIC_KEY);
+        config.setAdditionalAuthzParameters(Map.of("token_format", "jwt"));
```
Consider adding a test case here that checks whether the additional authz parameters are properly added to the call to the IdP, if possible.

strehle committed:
review

strehle committed:
add test

strehle committed:
cleanup

strehle committed:
test converage

strehle committed:
Feature: Allow sending static key/value pairs to the configured external IdP for authorization code flow

Add parameter additionalAuthzParameters (additional authorization parameters) to requests for auth code flow.
The usage is: OIDC IdPs have additional parameters, e.g. UAA itself has token_format to get special features.
The new parameter allows to customize the IdP with own parameters

strehle committed:
sonar

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feature/issue/2395

adrianhoelzl-sap commented on model/src/main/java/org/cloudfoundry/identity/uaa/provider/OIDCIdentityProviderDefinition.java:
```diff
@@ -90,6 +104,7 @@ public boolean equals(Object o) {
         if (this.passwordGrantEnabled != that.passwordGrantEnabled) return false;
         if (this.setForwardHeader != that.setForwardHeader) return false;
         if (this.jwtClientAuthentication != that.jwtClientAuthentication) return false;
+        if (this.additionalAuthzParameters != that.additionalAuthzParameters) return false;
```
Please use `equals` instead (also for the `jwtClientAuthentication` one line above)

strehle committed:
rebase

adrianhoelzl-sap commented on server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthIdentityProviderConfigValidator.java:
```diff
@@ -26,19 +31,27 @@ public void validate(AbstractIdentityProviderDefinition definition) {
-            if (!hasText(def.getTokenKey()) && def.getTokenKeyUrl() == null) {
-                errors.add("Either token key or token key URL must be specified");
+            if (oidcIdentityProviderDefinition.getAdditionalAuthzParameters() != null &&
+                oAuthStandardParameters.stream().anyMatch(oidcIdentityProviderDefinition.getAdditionalAuthzParameters().keySet()::contains)) {
```
Consider using the `isOAuthStandardParameter` method here.

strehle committed:
review

strehle committed:
review

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into feature/issue/2395

strehle committed:
review
